examples/graph/mpnn-molecular-graphs.py

```diff
@@ -0,0 +1,723 @@
+            carry_state_expanded = tf.gather(carry_state, atom_partition_indices)
+
+            # Perform a linear transformation followed by reduction
+            atom_features_reduced = tf.reduce_sum(
+                atom_features * carry_state_expanded, axis=1
+            )
+
+            # Split into parts correspoding to each molecule in the batch
+            atom_features_partitioned = tf.dynamic_partition(
+                atom_features_reduced, atom_partition_indices, self.batch_size
+            )
+
+            # Compute attention coefficients
+            attention_coef = tf.concat(
+                [tf.nn.softmax(features) for features in atom_features_partitioned],
+                axis=0,
+            )
+
+            # Apply attention to atom_features, and sum based on atom_partition_indices
+            attention_readout = tf.math.segment_sum(
+                tf.reshape(attention_coef, [-1, 1]) * atom_features,
+                atom_partition_indices,
+            )
+
+            # Concatenate (previous) carry_state and attention readout
+            carry_state_evolved = tf.concat([carry_state, attention_readout], axis=1)
+
+            # Perform a LSTM step (with only a memory state and carry state)
+            memory_state, carry_state = self.lstm_cell(
+                [memory_state, carry_state_evolved]
+            )
+
+        return carry_state_evolved
+
+
+"""
+### Message Passing Neural Network (MPNN)
+
+It is now time to complete the MPNN model. In addition to the message passing
+and readout, a two-layered classification network will be implemented to make
+predictions of BBBP.
+"""
+
+
+def MPNNModel(
+    atom_dim,
+    bond_dim,
+    batch_size=32,
+    message_units=64,
+    message_steps=4,
+    lstm_steps=8,
+    dense_units=256,
+):
+
+    atom_features = keras.layers.Input(
+        shape=(atom_dim), dtype="float32", name="atom_features"
+    )
+
+    bond_features = keras.layers.Input(
+        shape=(bond_dim), dtype="float32", name="bond_features"
+    )
+
+    pair_indices = keras.layers.Input(shape=(2), dtype="int32", name="pair_indices")
+
+    atom_partition_indices = keras.layers.Input(
+        shape=(), dtype="int32", name="atom_partition_indices"
+    )
+
+    x = MessagePassing(message_units, message_steps)(
+        [atom_features, bond_features, pair_indices]
+    )
+
+    x = SetGather(batch_size, lstm_steps)([x, atom_partition_indices])
+
+    x = keras.layers.Dense(dense_units, activation="relu")(x)
+
+    x = keras.layers.Dense(1, activation="sigmoid")(x)
+
+    model = keras.Model(
+        inputs=[atom_features, bond_features, pair_indices, atom_partition_indices],
+        outputs=[x],
+    )
+    return model
+
+
+mpnn = MPNNModel(
+    atom_dim=X_train[0][0][0].shape[0], bond_dim=X_train[1][0][0].shape[0],
+)
+
+mpnn.compile(
+    loss=keras.losses.BinaryCrossentropy(),
+    optimizer=keras.optimizers.Adam(learning_rate=5e-4),
```

Comment:
How did you obtain the learning rate?

Reply:
I first tried the default `1e-3` and it worked OK. However, lowering it to `5e-4` made the validation AUC (hence the learning curve) more stable.